src/script/iter.rs

```diff
@@ -0,0 +1,144 @@
+    mut iter: impl Iterator<Item = Result<T, script::Error>>,
+    flags: interpreter::Flags,
+    script_code: &script::Code,
+    stack: interpreter::Stack<Vec<u8>>,
+    checker: &dyn interpreter::SignatureChecker,
+) -> Result<interpreter::Stack<Vec<u8>>, script::Error> {
+    iter.try_fold(interpreter::State::initial(stack), |state, elem| {
+        elem.and_then(|op| {
+            op.eval(flags, script_code, checker, state)
+                .map_err(|e| script::Error::Interpreter(Some(op.clone().into()), e))
+        })
+    })
+    .and_then(|final_state| match final_state.vexec.len() {
+        0 => Ok(final_state.stack),
+        n => Err(script::Error::UnclosedConditional(n)),
+    })
+}
+
+/// Pre-version-0.6, Bitcoin always counted CHECKMULTISIGs as 20 sigops. With pay-to-script-hash,
+/// that changed: CHECKMULTISIGs serialized in script_sigs are counted more accurately, assuming
+/// they are of the form ... OP_N CHECKMULTISIG ...
+pub fn get_sig_op_count<T: Into<opcode::PossiblyBad>>(
+    iter: impl Iterator<Item = Result<T, opcode::Error>> + Copy,
+    accurate: bool,
+) -> u32 {
+    iter::once(Ok(None))
+        .chain(iter.map(|r| r.map(Some)))
+        .zip(iter)
+        .map(|ops| match ops {
+            (Ok(last_opcode), Ok(opcode)) => match opcode.into() {
+                opcode::PossiblyBad::Good(Opcode::Operation(op)) => match op {
+                    opcode::Operation::OP_CHECKSIG | opcode::Operation::OP_CHECKSIGVERIFY => 1,
+                    opcode::Operation::OP_CHECKMULTISIG
+                    | opcode::Operation::OP_CHECKMULTISIGVERIFY => {
+                        match last_opcode.map(|op| op.into()) {
+                            Some(opcode::PossiblyBad::Good(Opcode::PushValue(
+                                opcode::PushValue::SmallValue(sv),
+                            )))
+                                // Even with an accurate count, 0 keys is counted as 20 for some
+                                // reason.
+                                if accurate && opcode::push_value::SmallValue::OP_1 <= sv => {
+                                    u32::try_from(sv.to_num()).expect("`sv` is positive")
+                                }
+                            // Apparently it’s too much work to figure out if it’s one of the few
+                            // `LargeValue`s that’s valid, so we assume the worst.
+                            Some(_) => u32::from(interpreter::MAX_PUBKEY_COUNT),
+                            // We’re at the beginning of the script pubkey, so any pubkey count must
+                            // be part of the script sig – assume the worst.
+                            None => u32::from(interpreter::MAX_PUBKEY_COUNT),
+                        }
+                    }
+                    _ => 0,
+                },
+                _ => 0,
+            },
+            (_, _) => 0,
+        })
+        .sum()
```

Comment:
Possible bug: In the previous code, this returned `Err(n)`, short-circuiting the iterator (and its count) at the first bad opcode. In this code, we treat bad opcodes as 0, but keep counting.

If the previous behaviour was consensus-critical (which I suspect it was), change this back to using a mechanism with the early-return behaviour.